components/google_drive/sources/new-files/new-files.js

```diff
@@ -0,0 +1,227 @@
+        );
+        await this.googleDrive.stopNotifications(
+          channelID,
+          subscription.resourceId
+        );
+        const { expiration, resourceId } = await this.googleDrive.watchDrive(
+          channelID,
+          this.http.endpoint,
+          pageToken,
+          this.drive === "myDrive" ? null : this.drive
+        );
+        this.db.set("subscription", { expiration, resourceId });
+      }
+      return;
+    }
+
+    const { headers } = event;
+
+    if (headers["x-goog-resource-state"] === "sync") {
+      console.log("Sync notification, exiting early");
+      return;
+    }
+
+    if (
+      !headers["x-goog-resource-state"] ||
+      !headers["x-goog-resource-id"] ||
+      !headers["x-goog-resource-uri"] ||
+      !headers["x-goog-message-number"]
+    ) {
+      console.log("Request missing necessary headers: ", headers);
+      return;
+    }
+
+    const incomingChannelID = headers["x-goog-channel-id"];
+    if (incomingChannelID !== channelID) {
+      console.log(
+        `Channel ID of ${incomingChannelID} not equal to deployed component channel of ${channelID}`
+      );
+      return;
+    }
+
+    if (headers["x-goog-resource-id"] !== subscription.resourceId) {
+      console.log(
+        `Resource ID of ${resourceId} not currently being tracked. Exiting`
+      );
+      return;
+    }
+
+    // We observed false positives where a single change to a document would trigger two changes:
+    // one to "properties" and another to "content,properties". But changes to properties
+    // alone are legitimate, most users just won't want this source to emit in those cases.
+    // If x-goog-changed is _only_ set to "properties", only move on if the user set the prop
+    if (
+      !this.watchForPropertiesChanges &&
+      headers["x-goog-changed"] === "properties"
+    ) {
+      console.log(
+        "Change to properties only, which this component is set to ignore. Exiting"
+      );
+      return;
+    }
+
+    const {
+      changedFiles,
+      newStartPageToken,
+    } = await this.googleDrive.getChanges(
+      pageToken,
+      this.drive === "myDrive" ? null : this.drive
+    );
+
+    this.db.set("pageToken", newStartPageToken);
+
+    // only get files created since the last logged file creation or within the last five minutes
+    const fiveMinutesAgo = new Date(Date.now() - 5 * 60000);
+    const lastFileCreatedTime =
+      new Date(this.db.get("lastFileCreatedTime")) || fiveMinutesAgo;
+    let maxCreatedTime = lastFileCreatedTime;
+    let inFolder = false;
+    for (const file of changedFiles) {
+      const fileInfo = await this.googleDrive.getFile(file.id);
+      if (this.folders.includes("allFolders")) inFolder = true;
+      else {
+        for (const folder of fileInfo.parents) {
+          if (this.folders.includes(folder)) inFolder = true;
+        }
+      }
```

Comment:
Question: have you considered using this API instead?
https://developers.google.com/drive/api/v3/reference/files/watch

I haven't tried it myself, but if we could provide the folder ID's and watch those folders directly we could avoid this extra filtering, and only process events for the folders that we want. I'd imagine that a big shared drive with a lot of files can trigger a lot of irrelevant executions.

If not, I'd suggest we refactor this a bit so as to make it more performant:
```js
  methods: {
    shouldProcess(file) {
      // We should probably pre-compute this during `activate()`
      const watchedFolders = new Set(this.folders);

      return (
        watchedFolders.has("allFolders") ||
        file.parents.some(p => watchedFolders.has(p))
      );
    },
  },
  async run(event) {
    // ...
    for (const file of changedFiles) {
      const fileInfo = await this.googleDrive.getFile(file.id);
      if (shouldProcess(fileInfo)) {
        // Emit the event here...
      }
    }
  }
```

Reply:
I tried using files.watch with the folder id's instead of watching the drive, but it doesn't work for files in nested folders, and I wasn't able to get it to work for the "root" directory, so I used your refactoring suggestions.